### PR TITLE
fix unhandled transaction error

### PR DESCRIPTION
**Pull request by elpsyr**
Signed-off-by: crui <crui@relaper.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Fix some unhandled transaction error.
If the error caused by rollback is not captured, it will be commit successfully.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: edge/pkg/devicetwin/dtclient/device_db.go [hunk 1]
```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
...
 	}
 
 	for _, attr := range addAttrs {
 		err = SaveDeviceAttr(&attr)
 		if err != nil {
-			obm.Rollback()
+			errRollback := obm.Rollback()
+			if errRollback != nil {
+				klog.Errorf("transaction rollback failed: %v", errRollback)
+				return errRollback
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/device_db.go:117]: I think we should return the SaveDeviceAttr err,  it is the main err that why AddDeviceTrans exit, so  ignore the rollback err is better

File: edge/pkg/devicetwin/dtclient/device_db.go [hunk 2]
```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
...
+			}
 			return err
 		}
 	}
-	obm.Commit()
+	err = obm.Commit()
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/device_db.go:180]: suggest `return obm.Commit()`

File: edge/pkg/devicetwin/dtclient/device_db.go [hunk 3]
```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
...
+			}
 			return err
 		}
 	}
 
 	for _, twin := range addTwins {
 		err = SaveDeviceTwin(&twin)
 		if err != nil {
-			obm.Rollback()
+			errRollback := obm.Rollback()
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/device_db.go:126]: ditto

File: edge/pkg/devicetwin/dtclient/device_db.go [hunk 4]
```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
...
+	if err != nil {
+		klog.Errorf("begin transaction failed: %v", err)
+		return err
+	}
+	for _, deleteID := range deletes {
+		err = DeleteDeviceByID(deleteID)
 		if err != nil {
-			obm.Rollback()
+			errRollback := obm.Rollback()
+			if errRollback != nil {
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/device_db.go:155]: ditto

File: edge/pkg/devicetwin/dtclient/device_db.go [hunk 5]
```diff
@@ -89,58 +89,98 @@ func UpdateDeviceMulti(updates []DeviceUpdate) error {
...
+				return errRollback
+			}
 			return err
 		}
-		err = DeleteDeviceAttrByDeviceID(delete)
+		err = DeleteDeviceAttrByDeviceID(deleteID)
 		if err != nil {
-			obm.Rollback()
+			errRollback := obm.Rollback()
+			if errRollback != nil {
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/device_db.go:164]: ditto

File: edge/pkg/devicetwin/dtclient/deviceattr_db.go [hunk 6]
```diff
@@ -100,30 +100,50 @@ func UpdateDeviceAttrMulti(updates []DeviceAttrUpdate) error {
...
+			if errRollback != nil {
+				klog.Errorf("transaction rollback failed: %v", errRollback)
+				return errRollback
+			}
 			return err
 		}
 	}
-	obm.Commit()
+	err = obm.Commit()
+	if err != nil {
```
Comment (wackxu) [inline edge/pkg/devicetwin/dtclient/deviceattr_db.go:144]: ditto

File: edge/pkg/devicetwin/dtclient/device_db.go (hunk 1 above)
Comment (elpsyr) [reply edge/pkg/devicetwin/dtclient/device_db.go:117]: you are right，**no matter whether the rollback is successful or not**, the transaction will not be committed, so there is no need to handle the rollback error .
i will close this pr

**Comment by elpsyr** [comment]
no matter whether the rollback is successful or not, the transaction will not be committed, so there is no need to handle the rollback error .

**Comment by wackxu** [comment]
for begin or commit err, I think we should hanle it,  so other code lgtm

**Comment by elpsyr** [comment]
yeah,i will reopen it later

**Comment by elpsyr** [comment]
@wackxu I handle the error of transaction begin&commit,and leave some commit about rollback like this:
```
// we do not handle the error of rollback because:
// 1. no matter whether the rollback is successful or not,
// the transaction will not be committed.
// 2. err of SaveDevice is the main err that why AddDeviceTrans exit,
// so we should the SaveDevice err.
```
thanks  for your patient : )

**Comment by elpsyr** [comment]
btw ,sometimes i add the `Signed-off-by: Author Name <authoremail@example.com> ` in commit message,but i still fail with DCO check. 
Is there anything else to pay attention to?

**Pull request closed without merging**